File: Programmers/[92345] 사라지는 발판/JFe/92345.py
def solution(board, aloc, bloc):
    def dfs(ax, ay, bx, by):
        result = 0
        # 플레이어가 밟고 있는 발판이 사라졌다면
        if visited[ax][ay]: return 0
        # 상하좌우 탐색
        for d in range(4):
            next_x = ax + dx[d]
            next_y = ay + dy[d]
            if 0 <= next_x < r and 0 <= next_y < c:
                if visited[next_x][next_y] or board[next_x][next_y] == 0: continue
                visited[ax][ay] = True
                # 이동시켰을 때 턴의 수
                count = dfs(bx, by, next_x, next_y) + 1
                visited[ax][ay] = False
                # 현재 턴은 패배, 새로 계산된 턴은 승리인 경우
                if result % 2 == 0 and count % 2 == 1:
                    result = count
                # 현재 턴과 새로운 턴 모두 패배인 경우
                elif result % 2 == 0 and count % 2 == 0:
                    result = max(result, count)
                # 현재 턴과 새로운 턴 모두 승리인 경우
                elif result % 2 == 1 and count % 2 == 1:
                    result = min(result, count)
        return result
    r, c = len(board), len(board[0])
    visited = [[False for _ in range(5)] for _ in range(5)]
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    
    answer = dfs(aloc[0], aloc[1], bloc[0], bloc[1])
    return answer

File: Programmers/[92345] 사라지는 발판/ONE/92345.java
class Solution {
    class Result{
        boolean win;
        int count;

        public Result(boolean win, int count){
            this.win = win;
            this.count = count;
        }
    }

    private int r, c;
    private static int[][] dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    public int solution(int[][] board, int[] aloc, int[] bloc) {
        r = board.length;
        c = board[0].length;
        return dfs(aloc[0], aloc[1], bloc[0], bloc[1], 0, board).count;
    }

    public Result dfs(int curX, int curY, int anotherX, int anotherY, int count, int[][] board){
        boolean win = false;
        int minCount = 5 * 5;
        int maxCount = count;

        if(board[curX][curY] == 1){
            for(int[] tmp : dir){
                int x1Tmp = curX + tmp[0];
                int y1Tmp = curY + tmp[1];
                if(isValid(x1Tmp, y1Tmp, board)){
                    board[curX][curY] = 0;
                    Result d = dfs(anotherX, anotherY, x1Tmp, y1Tmp, count + 1, board);
                    win |= !d.win;
                    if (d.win)
                        maxCount = Math.max(maxCount, d.count);
                    else
                        minCount = Math.min(minCount, d.count);
                    board[curX][curY] = 1;
                }
            }
        }

        return new Result(win, win ? minCount : maxCount);
    }

    public boolean isValid(int x, int y, int[][] board){
        return x >= 0 && x < r && y >= 0 && y < c && board[x][y] != 0;
    }
}